contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,265 @@
+  execute the run tasks on a given volume or pool.
+- Administrator will have to create a Kubernetes Job out of the
+  upgrade agent - by passing the upgrade cas template and the object
+  details via a config map. 
+
+While the above approach was good enough to be automated via `kubectl`, 
+the steps that can be executed as part of upgrade were limited
+by the features or constructs available within the CAST/RunTasks. 
+
+Another short-comming of the CAST/RunTasks is the lack of developer
+friendly constructs for rapidly writing new CAST/RunTasks. This was
+severely impacting the time taken to push out new releases. 
+
+### Proposed Implementation
+
+Since the upgrades using the scripts have been well tested and tried 
+out on several test and production developments, we decided to re-use
+the shell scripts to perform the upgrade tasks. 
+
+The upgrade scripts will be packaged into upgrade container, with an 
+entrypoint script that will invoke the right scripts depending on 
+the object being upgraded. 
+
+A custom resource called `UpgradeTask` will be defined with the 
+details of the object being upgraded, and will in trun be updated
+with the status of the upgrade by the upgrade-container scripts.  
+
+OpenEBS team will release an openebs-upgrade-tools container 
+that contains all the upgrade scripts supported by openebs. 
+
+The workflow followed by administrator is as follows:
+- Define the Custom Resource and a Service account
+- Create a UpgradeTask CR for the object being upgraded 
+- Create a Kubernetes Job with upgrade-container and passing the 
+  UpgradeTask details. 
+- Administrator can query the UpgradeTask to check the status 
+  and result of the upgrade.
+
+### Design Choices/Decisions
+
+This section captures some of the alternative design considered
+and the reasoning behind selecting a certain approach. 
+
+- A generic UpgradeTask CR vs task specific ones like JivaUpgradeTask,
+  CStorVolumeUpgradeTask, etc. Having specific tasks has the following
+  advantages:
+  * Each tasks can have its own spec. the fields may vary depending
+    on the resource being upgraded.
+  * Writing specific operators for each upgrade that operates on a
+    given type of task will make the upgrades more modular. 
+  
+  However, this also means that every time a new resource type is
+  added, another CR needs to be introduced, managed and learned by 
+  the user. This may still be ok. But a similar pattern where 
+  different specs are required is already addressed by the PVC. To 
+  keep the resources management at a minimum, the PVC type, sub-resource
+  spec pattern will be used to specifiy different resources under
+  a generic UpgradeTask CR. 
+
+  Note that, selecting the Generic Task - doesn't preclude from 
+  writing specific upgrade task operators and upgrades task CRs as long
+  as the management of these specific upgrade tasks are completely 
+  management by the operators and user need not interact with them. One
```

Comment:
So, we are going with the approach to have a generic upgrade task.

Reply:
yes.